design/ceph/object/read-affinity.md

```diff
@@ -0,0 +1,92 @@
+- Endpoints get can get overloaded due to insufficient numbers of endpoints or too many requests. In this case, the RGW daemons on that zone can be scaled using KEDA.
+
+
+## Backend Data Locality
+- RGW Read Affinity can reduce cross-zone traffic by keeping the data access within the respective data centers.
+- For example, in a stretch cluster deployed in a multi-zone environment, the read affinity topology implementation helps to keep the traffic within the data center it originated from.
+- RGW daemons have the ability to read data from an OSD in promixity to the RGW client, according to the OSD location defined in the CRUSH map and the topology labels on the nodes.
+- Following Read Affinity options are available in ceph:
+    - `Localize`: Read from OSD that is nearest to the RGW client daemon that has received a GET request.
+    - `Balanced`: Read from a random OSD in the Placement Group's active set of OSDs.
+    - `Default`: Read from the Primary OSD.
+- `Localize` mode should be used in order to avoid cross-zone traffic.
+
+### Implementation:
+
+####  `API Changes`
+```yaml
+apiVersion: ceph.rook.io/v1alpha1
+kind: CephObjectStore
+metadata:
+  name: my-store
+  namespace: rook-ceph
+spec:
+  gateway:
+    readAffinnity:
+        type: localize
+        failureDomain: zone
+```
+Where:
+    - `type`: refers to the RGW read Affinity type. It can be `localize`, `balanced` or `default`. Use `Localize` reads to avoid cross-zone traffic.
+    - `failureDoman`: is the RGW daemon failure domain. The RGW deployments will be scaled across the failureDomain to ensure that there is at least one RGW daemon pod running on each failure domain.
+
+- Use Zone specific deployments to create a separate RGW deployment for each zone.
+
+#### Set Crush Location
+- Crush location should be set on the RGW daemon in order to determine the nearest OSD in `localize` mode.
+- Crush location can be specificed by passing  `--crush-location=host=<hostName> zone=<zoneName> root default` argument in in `radosgw` command in the RGW daemon pod.
+- Node name for the RGW is daemon is not known until the pod is scheduled. So its not easy to provide the node topolgy labels like  and use them in the `--crush-location` argument.
+- There is an k8s KEP to use node topology labels as backward API. Until that is implmmented, we should be relying on some other workaround to pass the labels into the RGW daemon POD container.
+- Some available workarounds:
+    - `Option 1: topology labels as configmMap`.
+        - All the node topology labels can be stored in config map as a json file.
+        - ConfigMap can be mounted on the RGW deployment as a file.
+        - `jq` utility can be used to parse this json file and set the node topolgy labels as environment variables.
+        - The topology labels set as environment variables can be parsed and used in the `--crush-location` argument.
+    - `Option 2: Canary pods for RGW daemons`
```

Comment:
Option 2 doesn't seem worth the effort, since we expect the downward API will solve this for us, when that feature is available.

Reply:
downward API won't/might not work for custom topologies.